backend/src/main/java/com/bakdata/conquery/models/preproc/Preprocessed.java

```diff
@@ -172,21 +176,74 @@ private Map<String, ColumnStore> combineStores(Object2IntMap<String> entityStart
 		return columnStores;
 	}
 
-	private static void writePreprocessed(File file, PreprocessedHeader header, PreprocessedData data) throws IOException {
+	private static void writePreprocessed(File file, PreprocessedHeader header, Map<String, Integer> globalStarts, Map<String, Integer> globalLengths, Map<String, ColumnStore> data) throws IOException {
 		final OutputStream out = new GZIPOutputStream(new FileOutputStream(file));
 		try (JsonGenerator generator = Jackson.BINARY_MAPPER.copy().enable(JsonGenerator.Feature.AUTO_CLOSE_TARGET).getFactory().createGenerator(out)) {
 
 			log.debug("Writing header");
 
 			generator.writeObject(header);
 
-
 			log.debug("Writing data");
 
-			generator.writeObject(data);
+			final Map<Integer, List<String>>
+					entity2Bucket =
+					globalStarts.keySet().stream().collect(Collectors.groupingBy(id -> Hashing.consistentHash(HashCode.fromString(id), 1000)));
+
+			for (Map.Entry<Integer, List<String>> bucketIds : entity2Bucket.entrySet()) {
+				final HashSet<String> entities = new HashSet<>(bucketIds.getValue());
+				final Map<String, Integer> starts = Maps.filterKeys(globalStarts, entities::contains);
+				final Map<String, Integer> lengths = Maps.filterKeys(globalLengths, entities::contains);
+
+				final PreprocessedData preprocessedData = selectBucket(bucketIds.getKey(), starts, lengths, data);
+
+				generator.writeObject(preprocessedData);
+			}
```

Comment:
Kommst du so nicht immer auf ~1000 Buckets mit beliebig vielen Entities anstatt 1000 Entities per Bucket?.

Reply:
Also ich verstehe die Idee, aber die 1000 sind ein schwierig zu tunender Parameter glaube ich, da so für einen großen Datensatz große Buckets und für einen kleinen Datensatz kleine Buckets heraus kommen, was wiederum für sehr unterschiedlich große Nachrichten zwischen Manager und Shards sorgt. Wir sollten versuchen die Nachrichten größe zu kappen (Spaltenanzahl kann uns da immer noch einen Strich durch die Rechnung machen).

Reply:
um wieder auf eine gekappte Bucketsize von 1000 zukommen  könnte man `entity2Bucket` pivoten.

Reply:
> Kommst du so nicht immer auf ~1000 Buckets mit beliebig vielen Entities anstatt 1000 Entities per Bucket?.

das ist wahr, wobei die gleichverteilt auf den keys sind. ich glaube es nimmt sich kaum was, außer wie du sagst, dass die größe der nachrichten fluktuiert. Das kann man sich noch anschauen; ich denke aber nicht, dass das so ein großes Problem ist, wir haben ja gestern schon gesehen, dass große Nachrichten gut durchs system gehen. 

Alternativ, ist das eine Zahl, die man beim einrichten eines Datensatzes überschlägt und dann festlegt und in einer config hinterlegt, die größe dürfte nicht so wichtig sein.